### ISSUE TITLE
storage: high memory usage of compactor

### ISSUE THREAD

**Issue by skyzh** [risingwavelabs/risingwave#3921]
<img width="929" alt="image" src="https://user-images.githubusercontent.com/4198311/179179625-95cc8f2f-0c5d-4191-bee0-bffe424c6711.png">

In TPC-H Q12 3-node benchmarks.

**Comment by xxhZs** [comment]
I find it too. And It doesn't happen when I do checkpoint every 5s

**Comment by jon-chuang** [comment]
After some offline discussion with Arne, here are some ideas:
1. Do not load all SST blocks into memory. 
2. Memory-limit k-way merge sort (funnel sort?). Needs multiple stages where intermediate results are written back to storage.
3. Add some "slack" to each SST so that merging into it will not result to a cascade of writes (probably does not help that much)
4. Interval scheduling of SSTs so that we do not have an all-or-nothing approach to merging if there are any non-overlapping ranges.

**Comment by skyzh** [comment]
The ideas look good to me. cc @hzxa21

**Comment by lmatz** [comment]
What's "slack"? 🤔

**Comment by ALeitert** [comment]
It was an idea I had when thinking about the tables and doing some test on my machine. I noticed that the compactor is often loading and rebuilding lots of tables and compacting them even if there is only a small change. And, as mentioned, a small change at the beginning, has a cascading effect to all following. Extreme case would be that we add one KV pair in the very first table. The idea is not fully developed yet, and I would also not give it priority. I also do not know how likely such cases are for real-life data.

**Comment by jon-chuang** [comment]
In terms of implementation complexity, @ALeitert points out that 4. it is as simple as organizing the SSTables to be compacted into groups of non-overlapping ranges. Then, we merge over each of these `ConcatIterator`s (which lazily load `SSTables` into memory) in the `MergeIterator`. 

If we have a small number of groups relative to total number of tables being merged, we would be successful in reducing the memory usage at any given point in time while improving our yield for large number of SSTables that have been merged into non-overlapping SSTs. So I would disagree about added complexity for 4.

---

However, we do need to validate if this helps our workloads. If our workload looks like merging a large number of non-overlapping SSTs with a small number of overlapping SSTs, then it will help. If it looks like merging a large number of overlapping ranges, then it will not help, and we need to rely on:
1. Idea 1 (stream blocks for `SSTableIterator` instead of loading them eagerly)
2. Limiting number of SSTs we merge at once
3. Reduce parallelism for compaction (not ideal?)

Arne is working on validating what kind of workload we have by measuring whether interval scheduling can result in small number of groups relative to number of sstables in the task and whether the maximum memory used for storing SSTables in a task at any point of time is reduced.

Generally, I think that 2 and 3 are poor compromises. We want a higher yield (large collections of SSTs with non-overlapping ranges per second) for fixed amount of memory. 2 results in fixed number of collections of SSTs, but with overlapping ranges. 3 results in fewer SSTs yielded per second.

---

As for 1. I'm not sure how good of an idea it is, as we can save costs from fetching a large contiguous SST from memory (single S3 get request). So I would be hesistant about the prefetch/stream approach.

---

The final idea that has not been covered so far is improving the way in which scheduling is performed. To my understanding, one possible ideal scenario is:

A small number N of `SSTable` with large range (RHS) are assigned to one task, to be merged with a large number M of non-overlapping `SSTables` each with small range (LHS). Then, the memory usage is N + 1 which is << N + M.

**Comment by hzxa21** [comment]
We check the compactor log and find out that the compaction task itself is not huge but the compactor node is running 6-7 compaction task concurrently. I think we should limit the number of in-flight compaction tasks per compactor node first.

**Comment by jon-chuang** [comment]
> compaction task itself is not huge but the compactor node is running 6-7 compaction tasks concurrently.

But I think above, we are trying to make each compaction task itself more efficient, allowing for more compacition tasks to be run concurrently on the same memory budget, resulting in higher SST yield.

**Comment by Little-Wallace** [comment]
> Interval scheduling of SSTs so that we do not have an all-or-nothing approach to merging if there are any non-overlapping ranges.
@jon-chuang 
I have implement it called `trivial move` in https://github.com/singularity-data/risingwave/pull/3111/files . If some SST does not overlap with anyone in the target level, it would be moved to the position in target level without any compaction task in compactor.

**Comment by jon-chuang** [comment]
I think rather than limit parallelism per se, we may be better off limiting the expected memory consumption. This way, we can set the target memory usage based on node's available memory, and we can even adjust dynamically according to realtime memory availability (assuming we are sharing the node with other processes). 

We can assume that N SSTs from L0 in the task results in N * L0 size bytes consumed, and any SST from L1 and above results in 1 * L? size bytes consumed.

**Comment by Little-Wallace** [comment]
In the origin implement we will read only one block and hold it in block-cache it means that we can control the memory of compactor precisely. But there is a problem that we must pay  AWS per IO requests. If one object is 32MB and the size of each of block is 64KB, it means we must pay 500 times money  than main branch.

**Comment by Little-Wallace** [comment]
> Actually, considering that currently, the largest SST file can be 2GB, streaming SSTable iter approach does seem reasonable. Perhaps, fetching blocks up to a buffer of 32MBs makes sense.

Exactly it could only reach 256MB and I will change it to 128MB. So I do not think it is the most important thing to reduce the memory usage for us. We only need to control the memory not exceed the upper limit and that is enough.
The cost of IO for S3 is much expensive.

**Comment by Little-Wallace** [comment]
🤔 In fact, I should prioritize identifying the various components of the memory occupy.

**Comment by hzxa21** [comment]
> Actually, considering that currently, the largest SST file can be 2GB, streaming SSTable iter approach does seem reasonable. Perhaps, fetching blocks up to a buffer of 32MBs makes sense.
> 
> However, note that those would be from levels with non-overlapping SST files. So we would have at most one large SST at a time. On the other hand, we store the SSTs resulting from compaction in memory for the duration of creation and upload as well. 
> 
> Since the target level is >= any of the inputs, the memory usage from keeping these files in memory instead of a streaming upload is already at least the size of largest input SST file. In other words, we may not be able to significantly reduce memory usage unless we also allow streaming upload of blocks of an SST prior to it being sealed...
> 
> Details:
> 
> S3 sdk requires a ByteStream implementing object, not necessarily `Bytes` like we provide now. So if we change the object store interface to accept some sort of byte stream instead of bytes, perhaps we can achieve a streaming upload. Specifically, in the S3 case, the way to achieve this would be to have `smithy::ByteStream::from(SdkBody::from(hyper::Body::channel().0)` as the upload parameter. 
> 
> In our case, our object store interface could have an `upload_stream(path, byte_stream)` interface where `byte_stream` is a generic bytestream implementing `Stream` trait. The s3 object store would take data from this bytestream and dump into the hyper byte stream (or simply use a `hyper::Body`'s `wrap_stream` interface). However, we should investigate Hyper's stream buffer size.
> 
> Likewise, object store should have a `read_stream` interface that allows reading `dyn Stream` object. Then, we can have a `BlockStream` and `SstableStream` objects that stream the bytes and produces the next block by calling `next().await`. etc.

@wenym1 has investigated streaming upload to S3 (see #1368) before. In short, it is not possible to achieve streaming upload via providing ByteStream to s3 sdk because s3 only supports http1. Yiming have plan to investigate multi-part upload in this Q and we will see how much it can help.


> We only need to control the memory not exceed the upper limit and that is enough

+1

**Comment by jon-chuang** [comment]
> In the origin implement we will read only one block and hold it in block-cache it means that we can control the memory of compactor precisely. But there is a problem that we must pay AWS per IO requests. If one object is 32MB and the size of each of block is 64KB, it means we must pay 500 times money than main branch.

Yes, I was thinking that single get request would be streamed over TCP, so we would pay for one get, but be able to stream the bytes of the object, pausing the stream while application buffer is full. Is this also not possible due to S3's use of HTTP1?

**Comment by jon-chuang** [comment]
> Exactly it could only reach 256MB and I will change it to 128MB.

Oh, hold on, I thought that 
L0 file size is
```rust
const DEFAULT_TARGET_FILE_SIZE_BASE: u64 = 32 * 1024 * 1024; // 32MB
```
While due to
```rust
input.target_file_size = self.config.target_file_size_base
    << (input.target_level.level_idx as usize - base_level);
```
L6 file size is 32MB * 2^6 = 2GB?

**Comment by jon-chuang** [comment]
> We only need to control the memory not exceed the upper limit and that is enough.

If we can achieve stream without additional S3 get request, I think we get a zero-cost memory usage optimization. 
.
But it may not be worth it if we cannot do streamed or multi-part upload to S3 as then we get at most factor of 2 reduction in memory usage, instead of possibly factor > 10.

**Comment by skyzh** [comment]
> S3 sdk requires a ByteStream implementing object, not necessarily Bytes like we provide now. So if we change the object store interface to accept some sort of byte stream instead of bytes, perhaps we can achieve a streaming upload. Specifically, in the S3 case, the way to achieve this would be to have smithy::ByteStream::from(SdkBody::from(hyper::Body::channel().0) as the upload parameter.

This requires us to provide body size in advance. So we still need to generate the full SST before uploading.

**Comment by jon-chuang** [comment]
> This requires us to provide body size in advance. So we still need to generate the full SST before uploading.

Yes, so I'm suggesting to use multipart upload API.

**Comment by wenym1** [comment]
> Yes, so I'm suggesting to use multipart upload API.

Sure I think we can do multipart upload. 

I had some rough idea before about providing a new upload API for object store to accept a stream like input and upload the whole SST portion by portion (ideally block by block). The similar idea will also be helpful to save memory for spill to disk, since a file handle can write like a stream instead of writing the whole object.

I plan to support this feature soon after I do some experiment about multipart upload on both S3 and MinIO. Besides, it seems that there is a limit on the minimum part size for multipart upload, so we need to handle the upload in a tricky way.

**Comment by jon-chuang** [comment]
> so we need to handle the upload in a tricky way.

I guess one requirement is that parts are uploaded sequentially after previous one is complete.

I think from our perspective, S3 object store should pull bytes from input stream. If input stream not exhausted, add to buffer of fixed size (min multipart part size). Then, if exhausted, just upload remainder. So its probably not too tricky.

**Comment by Little-Wallace** [comment]
> L6 file size is 32MB * 2^6 = 2GB?

No. in src/storage/src/hummock/compactor.rs  we can limit it by
```rust
        let target_file_size = std::cmp::min(
            self.compact_task.target_file_size as usize,
            max_target_file_size,
        );
```

**Comment by wenym1** [comment]
> I think from our perspective, object store interface implemented for S3 should pull bytes from input stream. If input stream not exhausted, add to buffer of fixed size (min multipart part size, or first multiple of block size greater than that). Once full, do upload of part. Else, if stream is exhausted, just upload remainder. So its probably not too tricky.

It's not as simple as this solution. 

When we upload the remainder, the remainder is also a part, but we cannot ensure that this part is greater than the minimum part size, which may cause the last upload part to fail (not sure about this yet, we may need some experiment on S3 and MinIO). A solution to this is, we always ensure the multipart upload buffer to have more than one part size data, and we will need a buffer of size of two times the minimum part size. When the buffer size exceeds two times the minimum part size, we upload the first minimum part size, and the remaining buffer size is still greater than the minimum part size. In this way, we can make sure the last part can be greater than the minimum part size. If the whole data is smaller than the minimum part size, just do a simple upload.

**Comment by jon-chuang** [comment]
> So I do not support uploading data by streaming API because it is not necessary.

Indeed, if we limit the size to ~100 MB, there is a reduced incentive to do streaming download or multipart upload, at least with saving memory as a motivation.

However, I guess we don't want to have too small file sizes, then we have to do more get requests for both metadata and data and maintain more metadata.

Further, there are other advantages to streaming: 
1. Performance - According to https://singularity-data.quip.com/RalDAVhdy5EI/S3-performance#eCFACA0rGWu, download of 128MB file takes on average 1.6s.
2. Upload 128MB takes 1.28s, and 0.69s for multipart.


If we can reduce memory cost significantly for 128MB * 10 input files, we can save ~ 1GB per task.

---

In terms of prioritizing streaming download v.s. streaming upload, given that when file size is large (~100MB), input file size == output file size, it may be worth spending more time optimizing the download rather than upload. Since for level multiplier = 10, input SSTs would be 10x output, the effectiveness of streaming the download is 10x of streaming the upload. So perhaps we should deprioritize streaming upload as it is also significantly more complicated (including storing output in block cache for local compaction).

**Comment by Little-Wallace** [comment]
@jon-chuang  I do not understand why you think the speed of download could be improved by 10x by streaming

**Comment by Little-Wallace** [comment]
The total size of input would be close to output in most of case.

**Comment by jon-chuang** [comment]
> I do not understand why you think the speed of download could be improved by 10x by streaming

Actually I meant memory consumption saved by streaming download would be 10x of streaming upload.

> The total size of input would be close to output in most of case.

Yes, but size of output SST == size of each input SST. Total held in memory = 10 X input + 1 X output.

**Comment by Little-Wallace** [comment]
The high memory usage of compactor seems to be caused by some bug. I think we shall fix bug before any optimize for read

**Comment by Little-Wallace** [comment]
> Yes, but size of output SST == size of each input SST. Total SST held in memory = 10 X input + 1 X output.
No. We only merge overlapping L0 no more than 128MB. It means that most cases, the input data onlys uses several tens MB.

**Comment by jon-chuang** [comment]
> No. We only merge overlapping L0 no more than 128MB. It means that most cases, the input data onlys uses several tens MB.

Don't we merge multiple Lk SSTs with single Lk+1 SST?

**Comment by jon-chuang** [comment]
Perhaps the issue is Arc<Block> in SSTable: https://github.com/singularity-data/risingwave/blob/main/src/storage/src/hummock/sstable/mod.rs

**Comment by jon-chuang** [comment]
You can see from an old implementation of load_table https://github.com/singularity-data/risingwave/blob/fadff1446325761deb2f06c3e187777b93a8f07e/src/storage/src/hummock/sstable_store.rs#L290 that the current usage is an abuse of the original design. meta_cache is only meant to store SSTables without data blocks. But here we are abusing it by storing data blocks. I believe this is the cause of high memory usage.

**Comment by jon-chuang** [comment]
To fix this, I believe we should post-fill a copy of the SSTable only after block-less one has been inserted into cache. We should also return an owned value of SSTable if possible.

Finally, we may want to consider reducing use of Arc, which is dangerous where memory leak is concerned, where possible.

**Comment by wenym1** [comment]
> Perhaps the issue is Arc<Block> in SSTable

The size of the blocks in `SSTable` is included when we calculate the cost of `SSTable` so the overall cache memory usage is not likely to be affected by including blocks in `SStable`, since when the LruCache reaches its capacity, it will evict some entries to ensure that its memory usage will not exceed the limit.

> Another possibility: load_table returns CachableEntry, which has an arc reference to LRU

Holding an Arc of the LruCache is necessary since we have to sure that the LruCache is not freed while the CachableEntry is accessing the pointer of LruHandle. 

> Btw, why does meta cache in SSTableStore not cache meta but actual SSTable?

The reason why @Little-Wallace chose to store the whole prefetched SST in `SSTable` is that, he found a previous misuse of `bytes::Bytes`. 

How we used to use `bytes::Bytes` is that we prefetched the whole SST and split the byte array into different owned block data `Bytes`, and add each block into the block cache. However, when we use the `slice` of `Bytes`, though it returns a new `Bytes` with static lifetime, it's not a memory copy, but instead a reference to the memory of the whole SST `Bytes`. Therefore, if we added such block data `Bytes`s into the block cache, the real SST memory will be freed until all blocks of the SSTs is evicted, which will increase the real memory usage of block cache. Therefore, @Little-Wallace chose to add the whole prefetched SSTs in `SSTable` as a temporary solution so that at least the real memory usage can be tracked correctly. I think he is working some solutions to make cache logic clearer.

> We are cacheing ALL sstables that are loaded! And after, we are not deleting the entries!

The `SSTable` will be evicted when the cache reaches its capacity, so as the real SST data it holds. So the entries will finally be deleted when SST is evicted.

**Comment by wenym1** [comment]
> Finally, we may want to consider reducing use of Arc, to reduce chance of memory leak (unrelated)

Memory leak only happens for Arc when cyclic dependency appears. The current Arc dependency graph is quite clearer and is acyclic, so I don't think it's a good idea to reduce the current use of Arc.

**Comment by wenym1** [comment]
Our current implementation of LruCache is not a strict limitation on memory usage. 

The externally referenced CacheEntry cannot be evicted. Therefore, when we insert a new entry and the cache is going to exceed the memory capacity, even after we have evicted all evict-able entry, the memory usage can still exceed the memory capacity. In this scenario, insertion to LruCache is still allowed instead of blocking the insertion or failing. The total memory usage will exceed the capacity for a while until some entries are not externally referenced and evicts its data in LruCache. 

I am not sure whether this increases the real memory usage of LruCache. We may focus on the metrics of the LruCache usage of compactor to see whether the LruCache is using more memory than capacity.

**Comment by Little-Wallace** [comment]
This bug is caused by the incorrectly charge for `Sstable` when it include data blocks. And I will fix it in https://github.com/singularity-data/risingwave/pull/4034/

**Comment by jon-chuang** [comment]
>  at least the real memory usage can be tracked correctly. 

Ok, but we should not pollute the meta cache to achieve this…?

**Comment by jon-chuang** [comment]
Anw, meta_cache is not really that useful in the context of compactor. Since compactor and compute do not share same sstable_store, and compactor will in general fetch a single SST once and then never again. For this reason, perhaps we can reduce the meta and block cache capacity significantly for compactor.

**Comment by Little-Wallace** [comment]
Compactor only use meta-cache. Because compactor always need the whole file, rather than some blocks.

**Comment by jon-chuang** [comment]
But do we even need a cache? We can just stream the blocks and have the stream do the prefetching. Cache seems like the wrong tool here.

**Comment by Little-Wallace** [comment]
This

**Comment by Little-Wallace** [comment]
I'm working on this issue by https://github.com/singularity-data/risingwave/pull/4590/
I will run some tests in next few days to see whether we can control the input memory.

**Comment by MrCroxx** [comment]
I did some experiments and thought there may be memory leak in compactor. I'm currently investigating it.
